packages/cli/src/lib/merge/README.md

```diff
@@ -0,0 +1,621 @@
+
+```js
+// a.js
+const x = { foo: 'foo' }
+export const y = [1, 2, 3, 4, 5]
+// b.js
+export const y = [5, 6, 7, 8, 9]
+const x = { bar: 'bar' }
+```
+
+Both `a.js` and `b.js` define the same symbols, but in different orders. The semantic identity of `x` and `y` are the same across both files, because they share:
+
+1. The same scope (module scope)
+2. The same name
+3. The same initializer type (object and array, respectively).
+4. The same exported-ness.
+
+Semantic identity also applies at arbitrary scope depth, and can look "inside" functions, objects, and arrays.
+
+```js
+// c.js
+const x = { foo: { bar: 'bar' } }
+const y = (a) => {
+  const z = [1, 2, 3]
+  return z.contains(a)
+}
+// d.js
+const y = (a) => {
+  const z = [4, 5, 6]
+  return z.contains(a)
+}
+const x = { foo: { baz: 'baz' } }
+```
+
+In these examples, `x`, `z` and the `ObjectExpression` to the right side of `foo:` have the same semantic identity. Intuitively, two nodes have the same semantic identity if you can trace the "path" from their scope up to the module scope, and find identical identifiers at each parent scope. Here are some counterexamples to help illustrate:
+
+```js
+// e.js
+const e = (a) => {
+  const z = [1, 2, 3]
+  return z.contains(a)
+}
+// f.js
+const f = (a) => {
+  // This `z` does not have the same semantic identity as e.z,
+  // because the names e and f differ.
+  const z = [1, 2, 3]
+  return z.contains(a)
+}
+```
+
+```js
+// g.js
+const g = {
+  foo: {
+    bar: {
+      baz : 'baz'
+    }
+  }
+}
+// g2.js
+const g = {
+  foo: {
+    bar: { // This object's ID is the same as g.foo.bar above,
+      blep: 'baz' // But this string's ID is not; g.foo.bar.blep vs. g.foo.bar.baz.
+    }
+  }
+}
+```
+
+The goal of semantic identifiers (roughly) is to produce a unique identifier for an AST node, such that iff two AST nodes with the same identifier appeared in the same file, a naming collision would occur. This has one important exception: import statements. Import statements are difficult to uniquely identify because the following is legal:
+
+```js
+import { foo } from 'src'
+import { bar } from 'src'
+```
+
+We have two choices here. Either we consider these semantically equivalent by comparing their source (`src`), or consider them semantically different by comparing their import specifiers (`foo` and `bar`). For the purpose of merging, we choose the former. It is more useful to compare their source, and consider these semantically equivalent. Merging import statements is generally tricky, because Javascript has a few unintuitive rules about how import statements may be structured. For example, you cannot mix a default namespace import with import specifiers, like so:
```

Comment:
```suggestion
We have two choices here. Either we consider these semantically equivalent by comparing their source (`src`), or consider them semantically different by comparing their import specifiers (`foo` and `bar`). For the purpose of merging, we choose the former. It is more useful to compare their source, and consider these semantically equivalent. Merging import statements is generally tricky, because JavaScript has a few unintuitive rules about how import statements may be structured. For example, you cannot mix a default namespace import with import specifiers, like so:
```